app/src/main/java/com/stevesoltys/seedvault/transport/backup/PackageService.kt

```diff
@@ -157,9 +157,20 @@ internal fun PackageInfo.isSystemApp(): Boolean {
     return applicationInfo.flags and FLAG_SYSTEM != 0
 }
 
-internal fun PackageInfo.allowsBackup(): Boolean {
+internal fun PackageInfo.allowsBackup(context: Context, userId: Int): Boolean {
     if (packageName == MAGIC_PACKAGE_MANAGER || applicationInfo == null) return false
-    return applicationInfo.flags and FLAG_ALLOW_BACKUP != 0
+
+    // This must be kept in sync with frameworks/base/services/backup/java/com/android/server/backup/utils/BackupEligibilityRules.java
+    val isSystemApp = applicationInfo.flags and FLAG_SYSTEM != 0
+    val ignoreAllowBackup = !isSystemApp
+        && (context.checkSelfPermission(Manifest.permission.LOG_COMPAT_CHANGE)
+        == PackageManager.PERMISSION_GRANTED)
+        && (context.checkSelfPermission(Manifest.permission.READ_COMPAT_CHANGE_CONFIG)
+        == PackageManager.PERMISSION_GRANTED)
+        && CompatChanges.isChangeEnabled(IGNORE_ALLOW_BACKUP_IN_D2D, packageName,
+        UserHandle.of(userId))
```

Comment:
Under what circumstances can this `IGNORE_ALLOW_BACKUP_IN_D2D` be false?

Reply:
https://review.calyxos.org/c/CalyxOS/platform_frameworks_base/+/10624/1/services/backup/java/com/android/server/backup/utils/BackupEligibilityRules.java

Reply:
Note that the above change was just to make testing easier, not planning to merge.

@Uldiniad when you move this to a method, it could be a comment.

Reply:
It would be nice if you could write a bit more prose instead just pasting links in the hope I can figure it out myself.

So what does this code mean? Is this something per-app? depending on the app's target sdk version? is it enabled automatically per app as soon as they upgrade their target?